test/functional/test_f_multi_keyring.py

```diff
@@ -0,0 +1,150 @@
+            key_provider=MasterKeyInfo(provider_id=_PROVIDER_ID, key_info=_KEY_ID),
+            data_key=b'*!\xa1"^-(\xf3\x105\x05i@B\xc2\xa2\xb7\xdd\xd5\xd5\xa9\xddm\xfae\xa8\\$\xf9d\x1e(',
+        ),
+        encrypted_data_keys=[
+            EncryptedDataKey(
+                key_provider=MasterKeyInfo(provider_id=_PROVIDER_ID, key_info=_KEY_ID),
+                encrypted_data_key=b"\xde^\x97\x7f\x84\xe9\x9e\x98\xd0\xe2\xf8\xd5\xcb\xe9\x7f.}\x87\x16,\x11n#\xc8p"
+                b"\xdb\xbf\x94\x86*Q\x06\xd2\xf5\xdah\x08\xa4p\x81\xf7\xf4G\x07FzE\xde",
+            )
+        ],
+        encryption_context=_ENCRYPTION_CONTEXT,
+        signing_key=_SIGNING_KEY,
+        keyring_trace=[
+            KeyringTrace(
+                wrapping_key=MasterKeyInfo(provider_id=_PROVIDER_ID, key_info=_KEY_ID),
+                flags={
+                    KeyringTraceFlag.WRAPPING_KEY_GENERATED_DATA_KEY,
+                    KeyringTraceFlag.WRAPPING_KEY_ENCRYPTED_DATA_KEY,
+                },
+            )
+        ],
+    ),
+]
+
+_MULTI_KEYRINGS = [
+    MultiKeyring(
+        generator=RawAESKeyring(
+            key_namespace=_PROVIDER_ID,
+            key_name=_KEY_ID,
+            wrapping_algorithm=WrappingAlgorithm.AES_256_GCM_IV12_TAG16_NO_PADDING,
+            wrapping_key=_WRAPPING_KEY_AES,
+        ),
+        children=[
+            RawRSAKeyring(
+                key_namespace=_PROVIDER_ID,
+                key_name=_KEY_ID,
+                wrapping_algorithm=WrappingAlgorithm.RSA_OAEP_SHA256_MGF1,
+                wrapping_key=rsa.generate_private_key(public_exponent=65537, key_size=2048, backend=default_backend()),
+            ),
+            RawRSAKeyring(
+                key_namespace=_PROVIDER_ID,
+                key_name=_KEY_ID,
+                wrapping_algorithm=WrappingAlgorithm.RSA_OAEP_SHA256_MGF1,
+                wrapping_key=rsa.generate_private_key(public_exponent=65537, key_size=2048, backend=default_backend()),
+            ),
+        ],
+    ),
+    MultiKeyring(
+        generator=RawRSAKeyring(
+            key_namespace=_PROVIDER_ID,
+            key_name=_KEY_ID,
+            wrapping_algorithm=WrappingAlgorithm.RSA_OAEP_SHA256_MGF1,
+            wrapping_key=rsa.generate_private_key(public_exponent=65537, key_size=2048, backend=default_backend()),
+        )
+    ),
+    MultiKeyring(
+        children=[
+            RawRSAKeyring(
+                key_namespace=_PROVIDER_ID,
+                key_name=_KEY_ID,
+                wrapping_algorithm=WrappingAlgorithm.RSA_OAEP_SHA256_MGF1,
+                wrapping_key=rsa.generate_private_key(public_exponent=65537, key_size=2048, backend=default_backend()),
+            ),
+            RawAESKeyring(
+                key_namespace=_PROVIDER_ID,
+                key_name=_KEY_ID,
+                wrapping_algorithm=WrappingAlgorithm.AES_128_GCM_IV12_TAG16_NO_PADDING,
+                wrapping_key=_WRAPPING_KEY_AES,
+            ),
+        ]
+    ),
+]
+
+
```

Comment:
Additional behaviors we need to test:

## constructor
- [ ] constructor called with a generator but no children must succeed
- [ ] constructor called with children but no generator must succeed
- [ ] constructor called a generator and no children must throw `TypeError`
- [ ] constructor called with a generator or any children that are not `Keyring`s must throw `TypeError`

## on_encrypt
- [ ] given no generator, `on_encrypt` with no `data_encryption_key` must throw `EncryptKeyError`
- [ ] given a keyring that does nothing (make one of these as a test util) as the generator, `on_encrypt` with no `data_encryption_key` must throw `EncryptKeyError` (or `GenerateKeyError` if you change to that)
- [ ] given a keyring that always adds an EDK (make one of these as a test util) in each of the generator and multiple children, the `on_encrypt` return value must contain `len(children) + (1 if generator else 0)` more encrypted data keys than the input materials (test this for each of: generator but no children, children but no generator, both children and generator)
- [ ] given a keyring that generates but does not add an EDK, and no children, the `on_encrypt` return value must contain exactly the number of EDKs as the input value (test with both no input EDKs and >0)
- [ ] given input materials that already contain a `data_encryption_key`, every keyring must be called (you can use mock keyrings for this to simplify checking)

## on_decrypt
- [ ] given decryption materials that already contain a `data_encryption_key`, we must not call any of the keyrings (you can use mock keyrings for this to simplify checking)
- [ ] given a keyring that never adds an EDK in each of the generator and mutliple children, every keyring must be called (you can use mock keyrings for this to simplify checking)
- [ ] given decryption materials that do *not* already contain a `data_encryption_key`, once a keyring that adds an EDK is encountered, we must not call any further keyrings (you can use mock keyrings for this to simplify checking)

One good starting point when thinking about what to test is to step through your code, and any time you encounter a different behavior, make sure that you are testing that. One metric to look at to start measuring what you have and have not tested is to look at the coverage report and look for lines *and logic branches* that are not covered.

Coverage is not a perfect metric, but it's a good starting point and a reasonably good negative metric: ie, good coverage numbers don't mean that you have complete tests, but bad coverage numbers do mean you don't have complete tests.